Add support for print reftests

jgraham opened the pull request.

jgraham committed:
Support multiple PNGs per reftest comparison

This paves the way for print reftests in which each page will be a
seperate image. At present none of the output formats actually support
multiple images, so we always have to pick a single image to actually
output.

jgraham committed:
Vendor pdf.js

This only takes the compiled files necessary for our use case (in
particular build/pdf.js and build/pdf.worker.js from the PDFjs
distribution

jgraham committed:
Fix support for StaticHandler without format_args

jgraham committed:
Add basic print reftest support to marionette executor

Adds support for print reftests, as set out in RFC 41 [1]. This is
based on the WebDriver Print endpoint, using pdf.js to render the
pages of the generated PDF back to PNGs for comparison. The pdf.js
process itself runs in the browser, so there's quite a lot of
back-and-forth traffic involved; this approach is not fast. But we're
unlikely to get too many print reftests, so it's a useful baseline.

 https://github.com/web-platform-tests/rfcs/pull/41

jgraham committed:
Add print-reftest support to the manifest

This adds a new test type to the manifest called print-reftest. Print
reftests are distinguished from normal reftests by filenames
containing -print as the type flag, or by being in directories called
'print'.

jgraham committed:
Add print reftest infrastructure test

jgraham committed:
Support print reftests in Chrome

This uses the underlying CDP command directly since PDF generation
isn't yet supported in Chrome. It also requires the browser to run in
headless mode since PDF generation is only supported in this mode.

gsnedders commented on infrastructure/reftest/reftest_match-print-ref.html:
```diff
@@ -0,0 +1,13 @@
+<meta name="reftest-pages" content="1-2">
+<style>
+* {margin: 0, padding:0}
+/* It is unclear to me why this is 180px rather than 192px (2 inches at 96dpi) */
+div {height:180px;}
```
Where does this 2in come from? But my first thought would be "what's the margin/padding on `@page`?".

jgraham replied:
Page is 5*3 inches with 0.5 inch margins.

stephenmcgruer replied:
Yeah I would try `@page { margin: 0; }` and see if that fixes it.

stephenmcgruer commented on tools/manifest/sourcefile.py:
```diff
@@ -410,6 +419,11 @@ def name_is_crashtest(self):
+    @property
+    def name_is_print_reftest(self):
+        # type: () -> bool
+        return self.type_flag == "print" or "print" in self.dir_path.split(os.path.sep)
```
We didn't spec a directory structure variant in the RFC, did we? Maybe an amendment to make.

stephenmcgruer commented on tools/wptrunner/wptrunner/browsers/chrome.py:
```diff
@@ -81,7 +83,8 @@ def executor_kwargs(test_type, server_config, cache_manager, run_info_data,
 
     # Pass the --headless flag to Chrome if WPT's own --headless flag was set
-    if kwargs["headless"] and "--headless" not in chrome_options["args"]:
+    if ((kwargs["headless"] or test_type == "print-reftest") and
```
Why do we need to put the browser into headless mode for print-reftests?

jgraham replied:
Don't know, you tell me ;) But it is apparently a limitation in Chrome that the `printToPDF` CDP endpoint only works in headless.

jgraham replied:
That sounds like a good question to ask a Blink engineer ;) (aiui it's a limitation of Chrome/CDP that the `printToPDF` endpoint only works in headless mode).

stephenmcgruer replied:
Ah, I see. Can you add a comment here noting that (maybe linking crbug.com/753118).

cc @foolip on this, since he's in the webdriver space these days and may know which devtools person to mention this to ;)

foolip replied:
I'm afraid is the first time I've seen crbug.com/753118. While I don't know more than what's on the bug, I've added it to a list of maybe good starter bugs for someone joining soonish.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -80,9 +80,10 @@ def __call__(self, test, result, extra=None):
 
 
-def hash_screenshot(data):
+def hash_screenshots(screenshots):
```
So our 'hash' is now a list of hashes; that's probably ok, but are we gaining anything from doing a hexdigest of each screenshot now or should we just return a list of base64 encodings?

Also nit, the docstring is now wrong.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -80,9 +80,10 @@ def __call__(self, test, result, extra=None):
+            for screenshot in screenshots]
 
 
 def _ensure_hash_in_reftest_screenshots(extra):
```
I haven't reviewed the rest of this file as it seems WIP - let me know if there's any part that needs looked at.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executorchrome.py:
```diff
@@ -1,9 +1,122 @@
+                "Loading initial page %s failed. Ensure that the "
+                "there are no other programs bound to this port and "
+                "that your firewall rules or network setup does not "
+                r"prevent access.\e%s" % (url, traceback.format_exc(e)))
```
I'm not familiar with `\e` here - what is it?

jgraham replied:
I'm thinking it's a typo that got introduced in 2015 and no one noticed until now.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executorchrome.py:
```diff
@@ -1,9 +1,122 @@
+        body = {
+            "cmd": "Page.printToPDF",
+            "params": {
+                "paperWidth": width / 2.54,
```
I assume then that the chrome-specific endpoint here only accepts inches, not cm? Worth documenting that if so, otherwise the `2.54` is sorta magical.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executorchrome.py:
```diff
@@ -1,9 +1,122 @@
+    def pdf_to_png(self, pdf_base64, ranges):
+        handle = self.webdriver.window_handle
+        self.webdriver.window_handle = self.runner_handle
+        print(self.webdriver.url)
```
Nit; leftover `print`

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executorchrome.py:
```diff
@@ -1,9 +1,122 @@
+            self.script = f.read()
+
+    def screenshot(self, test, viewport_size, dpi):
+        # https://github.com/web-platform-tests/wpt/issues/7135
```
Nit; https://github.com/web-platform-tests/wpt/issues/7140 is the dpi issue and even then we may want to track dpi support for print-reftests separately from reftests.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -879,7 +942,7 @@ def _screenshot(self, protocol, url, timeout):
             screenshot = screenshot.split(",", 1)[1]
 
-        return screenshot
+        return [screenshot]
```
It seems weird for a method to return a single item in a list, and looking at all the cases where `_screenshot` has been changed that seems to be what is happening. Assuming the caller is gathering a list, could it do the stitching instead?

Also, at least in executorwebdriver.py `_screenshot` wasn't changed, and I'm not sure if that was deliberate or not?

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -916,6 +979,9 @@ def run_test(self, test):
                                                               "timeout": timeout,
                                                               "width": 800,
                                                               "height": 600})["value"]
+        for item in rv["extra"].get("reftest_screenshots", []):
```
Again, same comment re wrapping a single-item list vs whoever is accumulating a list handling it.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -1016,3 +1082,60 @@ def do_crashtest(self, protocol, url, timeout):
+        self.protocol.pdf_print.load_runner()
+
+    def screenshot(self, test, viewport_size, dpi):
+        # https://github.com/web-platform-tests/wpt/issues/7135
```
Same comment re `dpi` issue

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executorwebdriver.py:
```diff
@@ -349,6 +349,7 @@ def run_func(self):
 
 class WebDriverTestharnessExecutor(TestharnessExecutor):
     supports_testdriver = True
+    protocol_cls = WebDriverProtocol
```
I don't understand what these changes are about; can you explain?

jgraham replied:
It's basically a mechanism to allow the `ChromeDriverPrintReftestExecutor` to inherit from the `WebDriverReftestExector` but also have a special `ChromeDriverProtocol` which is an extension of the `WebDriverProtocol` including the CDP-implemented print parts.

More abstractly it makes the protocol a swappable component rather that an intrinsic part of the executor.

stephenmcgruer commented on tools/wptrunner/wptrunner/wpttest.py:
```diff
@@ -554,6 +554,10 @@ def update_metadata(self, metadata):
             reference.update_metadata(metadata)
         return metadata
 
+    def get_viewport_size(self, override):
```
What is `override` for, if both implementations assert it is `None`?

jgraham replied:
It's basically just scafolding for the possibility of implementing this in the future. My guess is that different page sizes will be more important for paginated media than for visual.

stephenmcgruer replied:
Hrm, ok. I prefer not supporting code that isn't used, but I'll leave it up to you to decide.

stephenmcgruer commented on tools/wptrunner/wptrunner/wpttest.py:
```diff
@@ -589,6 +593,13 @@ def fuzzy_override(self):
+
+    def get_viewport_size(self, override):
+        assert override is None
+        return (5*2.54, 3*2.54)
```
So it seems like the webdriver API is in cm, whilst we spec'd this in inches. I presume most paginated media uses inches?

Any thoughts on converting here vs having `viewport_size` in inches and converting close to where we call webdriver?

jgraham replied:
I don't think "most paginated media uses inches"; all the A* paper sizes are specified in metric. AFAICT the use of inches by default is basically US-bias in the underlying API. I don't really want to use inches in the external facing parts.

stephenmcgruer replied:
I don't follow here - which underlying API are you referring to?

stephenmcgruer commented on tools/wptserve/wptserve/handlers.py:
```diff
@@ -467,6 +467,8 @@ def __init__(self, path, format_args, content_type, **headers):
         :param headers: List of headers to send with responses"""
 
         with open(path) as f:
-            data = f.read() % format_args
```
This change may not be wrong, but I'm not clear what it has to do with the rest of the PR?

jgraham replied:
Without this change we can't setup the print-reftest harness page without passing in empty format args. Note it's a separate commit in the series

stephenmcgruer reviewed (commented):
Looks fairly good at a high level, with some questions. Note that GitHub ate my initial review, so I may have missed some of my original comments when I retyped them all... (GitHubbbbbb! *shakes fist*)

jgraham committed:
Move wdspec detection higher in the order of precedence

This is primarilly so that print.py doesn't show up as a print reftest, but
also because wdspec tests are pretty easy to identify unambigously since they're
in a single subtree and are all python files.

jgraham committed:
Rename one WebDriverProtocol class as WdspecProtocol

This naming clash caused a confusing bug; really this class is only for wdspec tests.

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/executormarionette.py:
```diff
@@ -856,6 +856,11 @@ def __init__(self, logger, browser, server_config, timeout_multiplier=1,
+    def get_implementation(self, reftest_internal):
+        return (InternalRefTestImplementation if reftest_internal
+                else RefTestImplementation)(self)
+
```
Nit; should this be once whitespace-line between get_implementation and setup?

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -98,8 +98,57 @@ def _ensure_hash_in_reftest_screenshots(extra):
         if type(item) != dict:
             # Skip relation strings.
             continue
+        #TODO - Don't just take the final hash/screenshot, take a non-matching one
```
Highlighting a TODO (I'm sure you're aware of it) that I think should be resolved before this is merged.

stephenmcgruer commented on tools/manifest/sourcefile.py:
```diff
@@ -865,6 +865,16 @@ def manifest_items(self):
                     self.rel_path
                 )]
 
+        elif self.name_is_webdriver:
```
Oof, nice catch. Could we add a test so future-us don't get hurt when we refactor this?

stephenmcgruer reviewed (commented):
I think this is nearly there. Though I think I also reviewed this just before the latest force-push, not sure... so if any of my comments are weird I'm going to blame that and not my tiny brain :)

stephenmcgruer commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -370,28 +371,56 @@ def get_hash(self, test, viewport_size, dpi):
+    def check_pass(self, hashes, screenshots, urls, relation, fuzzy):
+        """Check if a test passes, and return a tuple of (pass, page_idx),
+        where page_idx is the zero-based index of the first page on which a
+        differene occurs if any, or None if there are no differences"""
```
```suggestion
        difference occurs if any, or None if there are no differences"""
```

stephenmcgruer commented on infrastructure/reftest/reftest_match-print-ref.html:
```diff
@@ -0,0 +1,13 @@
+<title></title>
+<meta name="reftest-pages" content="1-2">
+<style>
+* {margin: 0, padding:0}
```
Note that there's a comma here, is that valid as a CSS separator?

jgraham replied:
It's not, good catch. But fixing it doesn't seem to change anything.

jgraham committed:
Fix typo in docstring

Co-authored-by: Stephen McGruer <smcgruer@chromium.org>
